### PR TITLE
[Core] Recover MLS shape functions utility

### PR DESCRIPTION
**Pull request by rubenzorrilla**
This PR adds the capability to calculate MLS shape functions and gradients using a linear polynomial basis and an exponential kernel function.

The original implementation can be found in legacy applications repo ([here](https://github.com/KratosMultiphysics/KratosLegacyApplications/blob/master/LagrangianMPMApplication/custom_utilities/compute_mls_shape_functions_utility.h)). Changes with respect to the original implementation are basically the extension to the 3D case, optimization and testing.

### CODE REVIEW

File: kratos/utilities/mls_shape_functions_utility.cpp [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        const double h,
+        array_1d<double,TDim>& rKernelDerivative)
+    {
+        const double rad = norm_2(rRadVect);
+
+        if (rad < 1.0e-12) {
+            noalias(rKernelDerivative) = ZeroVector(TDim);
+        } else {
+            const double q = rad / h;
+            const double kernel_value = (1.0/(Globals::Pi*h*h)) * std::exp(-1.0*q*q) * (-2.0*q*(1.0/h));
```
Comment (RiccardoRossi) [inline kratos/utilities/mls_shape_functions_utility.cpp]: -q*q

and

-2.0*q/h

File: kratos/utilities/mls_shape_functions_utility.cpp [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+        // Set MLS shape functions containers
+        const std::size_t n_points = rPoints.size1();
+        if (rN.size() != n_points) {
+            rN.resize(n_points, false);
+        }
+
+        // Set the auxiliary arrays for the L2-norm problem minimization
+        static constexpr std::size_t BaseSize = TDim + 1;
+        std::vector<array_1d<double,BaseSize>> B_vect(n_points);
```
Comment (RiccardoRossi) [inline kratos/utilities/mls_shape_functions_utility.cpp]: do you need this to be an std::vector?

File: kratos/utilities/mls_shape_functions_utility.cpp [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+        // Set the polynomial basis y-derivative values at the point of interest
+        array_1d<double,3> Dp0_Dy;
+        Dp0_Dy[0] = 0;
+        Dp0_Dy[1] = 0;
+        Dp0_Dy[2] = 1;
+
+        // MLS shape function
+        array_1d<double,3> aux_prod;
+        for (std::size_t i_pt = 0; i_pt < n_points; ++i_pt) {
+            aux_prod = prod(M_inv, B_vect[i_pt]);
```
Comment (RiccardoRossi) [inline kratos/utilities/mls_shape_functions_utility.cpp:221]: noalias

File: kratos/utilities/mls_shape_functions_utility.cpp [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+        MathUtils<double>::InvertMatrix(M, M_inv, M_det);
+
+        // Set the polynomial basis values at the point of interest
+        array_1d<double,BaseSize> p0;
+        EvaluateLinearPolynomialBasis<TDim>(rX, p0);
+
+        // MLS shape function
+        array_1d<double,BaseSize> aux_prod;
+        for (std::size_t i_pt = 0; i_pt < n_points; ++i_pt) {
+            aux_prod = prod(M_inv, B_vect[i_pt]);
```
Comment (RiccardoRossi) [inline kratos/utilities/mls_shape_functions_utility.cpp]: noalias

File: kratos/utilities/mls_shape_functions_utility.cpp [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        const std::size_t n_points = rPoints.size1();
+        if (rN.size() != n_points) {
+            rN.resize(n_points, false);
+        }
+        if (rDNDX.size1() != n_points || rDNDX.size2() != 2) {
+            rDNDX.resize(n_points, 2, false);
+        }
+
+        // Set the auxiliary arrays for the L2-norm problem minimization
+        std::vector<array_1d<double,3>> B_vect(n_points);
```
Comment (RiccardoRossi) [inline kratos/utilities/mls_shape_functions_utility.cpp]: better if you use DenseVector instead

File: kratos/utilities/mls_shape_functions_utility.cpp [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+            const double kernel = MLSShapeFunctionsUtility::CalculateKernel(rad_vect, h);
+            MLSShapeFunctionsUtility::CalculateKernelDerivative<2>(rad_vect, h, w);
+
+            // Evaluate the current point basis
+            EvaluateLinearPolynomialBasis<2>(r_i_pt_coords, p);
+            noalias(p_outer_mat) = outer_prod(p,p);
+
+            // Add shape functions data
+            noalias(M) += kernel*p_outer_mat;
+            B_vect[i_pt] = kernel * p;
```
Comment (RiccardoRossi) [inline kratos/utilities/mls_shape_functions_utility.cpp]: noalias?

File: kratos/utilities/mls_shape_functions_utility.cpp [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+
+            // Add shape functions gradients data
+            noalias(DM_Dx) += w[0]*p_outer_mat;
+            noalias(DM_Dy) += w[1]*p_outer_mat;
+            DB_Dx_vect[i_pt] = w[0]*p;
```
Comment (RiccardoRossi) [inline kratos/utilities/mls_shape_functions_utility.cpp]: noalias?

File: kratos/utilities/mls_shape_functions_utility.cpp [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+        array_1d<double,4> Dp0_Dz;
+        Dp0_Dz[0] = 0;
+        Dp0_Dz[1] = 0;
+        Dp0_Dz[2] = 0;
+        Dp0_Dz[3] = 1;
+
+        // MLS shape function
+        array_1d<double,4> aux_prod;
+        for (std::size_t i_pt = 0; i_pt < n_points; ++i_pt) {
+            aux_prod = prod(M_inv, B_vect[i_pt]);
```
Comment (RiccardoRossi) [inline kratos/utilities/mls_shape_functions_utility.cpp]: noalias

**Review by RiccardoRossi — COMMENTED** [review]
minor performance comments.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp]: Leftover from the original implementation. Changing.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 8 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp]: Done.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 6 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp]: Done.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 8 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp]: Done.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 7 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp]: Done.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 4 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp]: Done.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 3 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp:221]: Done.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 5 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp]: Done.

File: kratos/utilities/mls_shape_functions_utility.cpp (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.cpp]: Not really. Done.

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**